### PR TITLE
PYIC-6437 Update EVCS stub, unable to pass query parameter "state" to GET api

### PR DESCRIPTION
**Pull request by rpayal**
## Proposed changes
Update EVCS stub, unable to pass query parameter "state" to GET api

### What changed
- Update EVCS stub for get api
- Change made to validate userId request param against access token 'sub' claim
- Updated openapi

### Issue tracking

- [PYI-6437](https://govukverify.atlassian.net/browse/PYI-6437)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts [hunk 1]
```diff
@@ -147,6 +158,24 @@ function isValidUpdateVcState(patchRequest: PatchRequest[]): boolean {
   return true;
 }
 
+function getRequestedStates(event: APIGatewayProxyEvent): string[] {
+  const STATE_ALL = "ALL";
+  const stateInQuery = event.queryStringParameters?.state;
+  let requestStates = Array.from((stateInQuery || VcState.CURRENT).split(","));
+  if (requestStates.length >= 1) {
```
Comment (Wynndow) [inline di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]: Do you need this check for if > 1? You could just run the `forEach` regardless?

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts [hunk 2]
```diff
@@ -147,6 +158,24 @@ function isValidUpdateVcState(patchRequest: PatchRequest[]): boolean {
...
+    requestStates.forEach(function (state) {
+      if (state != STATE_ALL && !(state in VcState)) {
+        throw new Error("Invalid state in query param.");
+      }
+    });
+  }
+  if (requestStates.length == 1 && requestStates.includes(STATE_ALL)) {
```
Comment (Wynndow) [inline di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts:170]: Could this just be `if (requestStates.includes(STATE_ALL))`?
If ALL is included, we don't really care if other states have been requested too?

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts [hunk 3]
```diff
@@ -165,9 +194,18 @@ function validateAccessToken(authheader: string | undefined): string {
...
 
-async function verifyAccessToken(jwt: string): Promise<boolean> {
+async function verifyAccessToken(
+  jwt: string,
+  userId: string,
+): Promise<boolean> {
   try {
-    return await verifyToken(jwt);
+    const payload: JWTPayload = await verifyTokenAndReturnPayload(jwt);
+    if (payload && payload.sub && userId !== payload.sub) {
```
Comment (Wynndow) [inline di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]:
```suggestion
    if (userId !== payload?.sub) {
```

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts [hunk 4]
```diff
@@ -165,9 +194,18 @@ function validateAccessToken(authheader: string | undefined): string {
...
+      throw new Error(
+        "User id doesn't match with `sub` claim value provided in the bearer token",
+      );
```
Comment (Wynndow) [inline di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]: This thrown error is getting caught in the catch block below, and then a new error thrown, which is a bit confusing. Maybe wrap the call to `verifyTokenAndReturnPayload` in the original try/catch and then do the userId/payload check and throw separately?

File: di-ipv-evcs-stub/lambdas/src/services/evcsService.ts [hunk 5]
```diff
@@ -76,23 +77,33 @@ export async function processPostUserVCsRequest(
...
+  const filterExpressions: string[] = [];
+  const ExpressionAttributeValues: Record<string, AttributeValue> = {};
+  ExpressionAttributeValues[`:userIdValue`] = { S: userId };
+  requestedStates.forEach((data, index) => {
+    filterExpressions.push(`:state${index}`);
+    ExpressionAttributeValues[`:state${index}`] = { S: data };
+  });
+  const FilterExpression = `#state IN (${filterExpressions.map(
+    (data) => data,
+  )})`;
```
Comment (Wynndow) [inline di-ipv-evcs-stub/lambdas/src/services/evcsService.ts]:
```suggestion
  const FilterExpression = `#state IN (${filterExpressions})`;
```

File: di-ipv-evcs-stub/openAPI/evcs-external.yaml [hunk 6]
```diff
@@ -115,19 +112,43 @@ paths:
...
           content:
             application/json:
               schema:
-                type: "object"
+                type: object
+                properties:
+                  vcs:
+                    type: string
+                    description: The SQS message ID.
+                    example: "bd8359d9-d559-47dd-9467-2a31e88a9e2d"
```
Comment (Wynndow) [inline di-ipv-evcs-stub/openAPI/evcs-external.yaml]: Thing this should be type array, not an SQS message ID

Comment (rpayal) [reply di-ipv-evcs-stub/openAPI/evcs-external.yaml]: yeah missed it.

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts (hunk 2 above)
Comment (rpayal) [reply di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts:170]: Want them to pass just 'ALL' .....test written for that.
valid;
- "ALL"
- "CURRENT,  PENDING"
invalid;
- "ALL, APPROVED"

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts (hunk 4 above)
Comment (rpayal) [reply di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]: done

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts (hunk 1 above)
Comment (rpayal) [reply di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]: done

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts (hunk 4 above)
Comment (rpayal) [reply di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts]: done

File: di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts (hunk 2 above)
Comment (Wynndow) [reply di-ipv-evcs-stub/lambdas/src/handlers/evcsHandler.ts:170]: I can't see anything in the spec for the actual evcs that adds that limitation - it's probably fine for now though

**Review by Wynndow — APPROVED** [review]

**Pull request merged**